nameko/amqp.py

```diff
@@ -47,3 +51,33 @@ def verify_amqp_uri(amqp_uri):
     transport = TestTransport(connection.transport.client)
     with transport.establish_connection():
         pass
+
+
+class UndeliverableMessage(Exception):
+    """ Raised when publisher confirms are enabled and a message could not
+    be routed or persisted """
+    pass
+
+
+@contextmanager
+def get_connection(amqp_uri):
+    conn = Connection(amqp_uri)
+    with connections[conn].acquire(block=True) as connection:
+        yield connection
+
+
+@contextmanager
+def get_producer(amqp_uri, confirms=True):
+    transport_options = {
+        'confirm_publish': confirms
+    }
+    conn = Connection(amqp_uri, transport_options=transport_options)
+
+    with producers[conn].acquire(block=True) as producer:
+        yield producer
+        try:
+            returned = producer.channel.returned_messages.get_nowait()
+        except Empty:
+            pass
+        else:
+            raise UndeliverableMessage(returned)
```

Comment:
It only raises when the `mandatory` flag is set on the message. It would be unusual to set mandatory delivery for an event dispatcher, but less so for a normal publisher. And reasonable to raise in any case if it is set.